### PR TITLE
Tlsv1.3

### PR DESCRIPTION
**Pull request by user-none**
Support for TLSv1.3 with OpenSSL 1.1.1.

### CODE REVIEW

File: tls/m_tls.c [hunk 1]
```diff
@@ -924,6 +930,38 @@ static M_bool M_io_tls_disconnect_cb(M_io_layer_t *layer)
...
+		return;
+
+	session = SSL_get1_session(handle->ssl);
+#if OPENSSL_VERSION_NUMBER >= 0x1010100fL && !defined(LIBRESSL_VERSION_NUMBER)
+	/* If it's not resumable we won't store it and
+ 	 * we want to remove any stored sessions from the
+	 * cache for this host and port because it shouldn't
+	 * be resumable either. */
+	if (session != NULL && !SSL_SESSION_is_resumable(session))
+		session = NULL;
```
Comment (bradh352) [inline tls/m_tls.c]: memory leak if, need to decrement reference counter as SSL_get1_session() incremented it

Comment (user-none) [reply tls/m_tls.c]: Fixed in 00b732f.

File: tls/m_tls_ctx_common.c [hunk 2]
```diff
@@ -39,11 +39,14 @@
 #include "base/m_defs_int.h"
 #include "m_tls_ctx_common.h"
 
+/* We'll tack these on at the end of the client and server ciphers. Our cipher setting funcion
+ * will parse out the TLSv1.3 only ones and non-TLSv1.3 from the string and set everything correctly. */
+#define TLS_v1_3_CIPHERS "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256"
 /* NOTE: Client still includes RC4 for now due to many servers only supporting RC4 due to
```
Comment (bradh352) [inline tls/m_tls_ctx_common.c]: Wonder if CHACHA20_POLY1305 should be first in the list, and the recommendation is that AES_128 shouldn't even be in the list it appears, or at least ssllabs downvotes it https://github.com/ssllabs/research/wiki/SSL-Server-Rating-Guide#cipher-strength  https://github.com/ssllabs/ssllabs-scan/issues/636

Comment (user-none) [reply tls/m_tls_ctx_common.c]: The list is the defaults OpenSSL currently sets. I wasn't sure if one is better than the other so I used their set. They only support 5 ciphers and default to those 3 on. Not a lot to choose from and I'd hope something so new would only use strong chipers.

**Comment by bradh352** [comment]
The only other thing is that TLS 1.3 servers can send back more than 1 'ticket' to use for future sessions, infact, OpenSSL defaults to '2':
https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_num_tickets.html

But with the current session code (inherited from the old code), it only grabs the current session handle on destruction of the connection for reuse rather than using the recommended callback approach to save each session ticket returned from the server:
https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_session_ticket_cb.html
So we'd have to store multiple sessions per host/port combo if we went that approach.

Also, best practice is to only use each session once, so we would actually delete the session once we used it:
https://wiki.openssl.org/index.php/TLS1.3#Sessions
Infact, it states for each session resumed, the server can (and openssl will) send yet another new ticket to use for a future session, so the callback will be called again.

**Comment by user-none** [comment]
> The only other thing is that TLS 1.3 servers can send back more than 1 'ticket' to use for future sessions, ...
> 
> But with the current session code (inherited from the old code), it only grabs the current session handle on destruction of the connection for reuse rather than using the recommended callback approach to save each session ticket returned from the server:

The callback approach is where we should go in the future. I don't think we need to go that far right now. The biggest reason for using the callback with TLSv1.3 is when storing the session during handshke. Since we store during shutdown we are getting a resumable sessions (there is a check if it really is resumable). I've confirmed the current code does work. I'd like to push changing the session storage to the future.

Also, I'm not sure we need to save multiple tickets. Looking at CURL (https://github.com/curl/curl/pull/3271/files) who switched from storing during handshake to the callback method, they only store the last ticket received and disgard any earlier ones.

> Also, best practice is to only use each session once, so we would actually delete the session once we used it:

The recommendation is to use it only once. However, it's not enforced and technically allowed to reuse sessions. I think that we should only allow sessions resuse once across all TLS versions and not restrict to only v1.3. It would also make it simipler to code, otherwise we'd have to hook into the handshake sequence to determine the TLS version to know if we should remove the session from the cache after setting it.

File: tls/m_tls_ctx_common.c (hunk 2 above)
Comment (user-none) [reply tls/m_tls_ctx_common.c]: > Wonder if CHACHA20_POLY1305 should be first in the list, and the recommendation is that AES_128 shouldn't even be in the list it appears, or at least ssllabs downvotes

Looking at the output of SSLabs server test against Google (who supports TLSv1.3) the v1.3 AES_128 isn't penalized. It's only penalizing <= v1.2 for that algorithm. I don't think we need to change from the OpenSSL defaults for security reasons.

**Pull request merged**